test/edgehog_device_forwarder/websockets_test.exs

```diff
@@ -0,0 +1,134 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.WebSocketsTest do
+  use EdgehogDeviceForwarder.CacheCase
+  use EdgehogDeviceForwarder.MessageCase
+  alias EdgehogDeviceForwarderProto.Edgehog.Device.Forwarder.Message, as: ProtoMessage
+  alias EdgehogDeviceForwarderProto.Edgehog.Device.Forwarder.WebSocket, as: ProtoWebSocket
+  alias EdgehogDeviceForwarder.{HTTPRequests, WebSockets}
+  alias EdgehogDeviceForwarder.WebSockets.Data
+
+  require Logger
+
+  setup_all do
+    [valid_message: {:text, "1"}]
+  end
+
+  setup %{
+    socket: controller,
+    http_upgrade_response: upgrade
+  } do
+    me = self()
+
+    request_id = HTTPRequests.new(controller)
+    socket = spawn(fn -> user_socket() end)
+
+    data = %Data{socket_id: request_id, device: me}
+    :ok = HTTPRequests.forward(request_id, upgrade)
+    :ok = WebSockets.upgrade(request_id, socket, data)
+
+    [
+      socket_id: request_id,
+      device_socket: me,
+      socket_data: data,
+      user_socket: socket
+    ]
+  end
+
+  describe "upgrade/3" do
+    setup %{socket: controller} do
+      request_id = HTTPRequests.new(controller)
+      upgrade_data = %Data{socket_id: request_id, device: self()}
+
+      [http_request_id: request_id, upgrade_data: upgrade_data]
+    end
+
+    test "does not allow an active http request to be upgraded", %{
+      http_request_id: request_id,
+      socket_data: data,
+      socket: websocket_socket
+    } do
+      assert {:error, :http_request_not_found} ==
+               WebSockets.upgrade(request_id, websocket_socket, data)
+    end
```

Comment:
Given the error atom, I would expect a different test description

Reply:
changed the description to `"returns an error if it called without first forwarding an upgrade request"`